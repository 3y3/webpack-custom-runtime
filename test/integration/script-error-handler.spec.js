'use strict';

const cct = require('./utils/custom-configurable-test');
const WebpackCustomRuntime = require('../../').CustomRuntimePlugin;

class TestPlainErrorHandler extends WebpackCustomRuntime.ErrorHandler {

    codeResolver(options, chunk, hash, { originalError, result }) {
        return `console.log(${originalError}, ${result});`;
    }
}

class TestFuncErrorHandler extends WebpackCustomRuntime.ErrorHandler {

    codeResolver() {
        return function(p1, p2, p3, p4) {
            console.log(p3, p4);
        }
    }
}

describe('scriptErrorHandler', function() {

    function cit(name, plugins = []) {
        cct(name, {
            plugins: [
                new WebpackCustomRuntime({ behavior: '' }),
                ...plugins
            ]
        });
    }

    cit('should reject original error, if there is no strategies');

    cit('should apply custom plain error handler', [
        new TestPlainErrorHandler()
    ]);

    cit('should apply custom func error handler', [
        new TestFuncErrorHandler()
    ]);

    cit('should apply multy error handlers', [
        new TestPlainErrorHandler(),
        new TestFuncErrorHandler()
    ]);
});
