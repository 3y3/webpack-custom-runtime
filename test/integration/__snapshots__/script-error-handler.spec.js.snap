// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scriptErrorHandler should apply custom func error handler: 0.js 1`] = `
"Index: 0.js
===================================================================
--- 0.js	
+++ 0.js	
"
`;

exports[`scriptErrorHandler should apply custom func error handler: 1.js 1`] = `
"Index: 1.js
===================================================================
--- 1.js	
+++ 1.js	
"
`;

exports[`scriptErrorHandler should apply custom func error handler: main-module.js 1`] = `
"Index: main-module.js
===================================================================
--- main-module.js	
+++ main-module.js	
"
`;

exports[`scriptErrorHandler should apply custom func error handler: other-module.js 1`] = `
"Index: other-module.js
===================================================================
--- other-module.js	
+++ other-module.js	
"
`;

exports[`scriptErrorHandler should apply custom func error handler: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ -161,1 +161,13 @@
-/******/ 	  return Promise.reject(originalError);
+/******/ 	  return [
+/******/ 	    function(p1, p2, p3, p4) {
+/******/ 	      console.log(p3, p4);
+/******/ 	    }
+/******/ 	  ]
+/******/ 	    .reduce(function(promise, strategy) {
+/******/ 	      return promise.then(function(result) {
+/******/ 	        return strategy(installedChunks, chunkId, originalError, result);
+/******/ 	      });
+/******/ 	    }, Promise.resolve(originalError))
+/******/ 	    .then(function(result) {
+/******/ 	      return result ? Promise.reject(result) : Promise.resolve();
+/******/ 	    });
"
`;

exports[`scriptErrorHandler should apply custom plain error handler: 0.js 1`] = `
"Index: 0.js
===================================================================
--- 0.js	
+++ 0.js	
"
`;

exports[`scriptErrorHandler should apply custom plain error handler: 1.js 1`] = `
"Index: 1.js
===================================================================
--- 1.js	
+++ 1.js	
"
`;

exports[`scriptErrorHandler should apply custom plain error handler: main-module.js 1`] = `
"Index: main-module.js
===================================================================
--- main-module.js	
+++ main-module.js	
"
`;

exports[`scriptErrorHandler should apply custom plain error handler: other-module.js 1`] = `
"Index: other-module.js
===================================================================
--- other-module.js	
+++ other-module.js	
"
`;

exports[`scriptErrorHandler should apply custom plain error handler: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ -161,1 +161,13 @@
-/******/ 	  return Promise.reject(originalError);
+/******/ 	  return [
+/******/ 	    function(installedChunks, chunkId, originalError, result) {
+/******/ 	      console.log(originalError, result);
+/******/ 	    }
+/******/ 	  ]
+/******/ 	    .reduce(function(promise, strategy) {
+/******/ 	      return promise.then(function(result) {
+/******/ 	        return strategy(installedChunks, chunkId, originalError, result);
+/******/ 	      });
+/******/ 	    }, Promise.resolve(originalError))
+/******/ 	    .then(function(result) {
+/******/ 	      return result ? Promise.reject(result) : Promise.resolve();
+/******/ 	    });
"
`;

exports[`scriptErrorHandler should apply multy error handlers: 0.js 1`] = `
"Index: 0.js
===================================================================
--- 0.js	
+++ 0.js	
"
`;

exports[`scriptErrorHandler should apply multy error handlers: 1.js 1`] = `
"Index: 1.js
===================================================================
--- 1.js	
+++ 1.js	
"
`;

exports[`scriptErrorHandler should apply multy error handlers: main-module.js 1`] = `
"Index: main-module.js
===================================================================
--- main-module.js	
+++ main-module.js	
"
`;

exports[`scriptErrorHandler should apply multy error handlers: other-module.js 1`] = `
"Index: other-module.js
===================================================================
--- other-module.js	
+++ other-module.js	
"
`;

exports[`scriptErrorHandler should apply multy error handlers: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ -161,1 +161,16 @@
-/******/ 	  return Promise.reject(originalError);
+/******/ 	  return [
+/******/ 	    function(installedChunks, chunkId, originalError, result) {
+/******/ 	      console.log(originalError, result);
+/******/ 	    },
+/******/ 	    function(p1, p2, p3, p4) {
+/******/ 	      console.log(p3, p4);
+/******/ 	    }
+/******/ 	  ]
+/******/ 	    .reduce(function(promise, strategy) {
+/******/ 	      return promise.then(function(result) {
+/******/ 	        return strategy(installedChunks, chunkId, originalError, result);
+/******/ 	      });
+/******/ 	    }, Promise.resolve(originalError))
+/******/ 	    .then(function(result) {
+/******/ 	      return result ? Promise.reject(result) : Promise.resolve();
+/******/ 	    });
"
`;

exports[`scriptErrorHandler should reject original error, if there is no strategies: 0.js 1`] = `
"Index: 0.js
===================================================================
--- 0.js	
+++ 0.js	
"
`;

exports[`scriptErrorHandler should reject original error, if there is no strategies: 1.js 1`] = `
"Index: 1.js
===================================================================
--- 1.js	
+++ 1.js	
"
`;

exports[`scriptErrorHandler should reject original error, if there is no strategies: main-module.js 1`] = `
"Index: main-module.js
===================================================================
--- main-module.js	
+++ main-module.js	
"
`;

exports[`scriptErrorHandler should reject original error, if there is no strategies: other-module.js 1`] = `
"Index: other-module.js
===================================================================
--- other-module.js	
+++ other-module.js	
"
`;

exports[`scriptErrorHandler should reject original error, if there is no strategies: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
"
`;
